Replace Generic Parameters with used Type Arguments in descriptions impl.

AdmiralSnyder opened the pull request:
this implements #41353

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=41641) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=41641) before we can accept your contribution.<br/><br/>:x: AdmiralSnyder [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=41641)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=41641) it.</sub>

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -269,6 +272,62 @@ private void EmitPendingChars()
+            }
+        }
+
+
```
delete all but one of these blank lines.

CyrusNajmabadi commented on src/Features/Core/Portable/Shared/Extensions/ISymbolExtensions_2.cs:
```diff
@@ -176,49 +176,8 @@ public static Glyph GetGlyph(this ISymbol symbol)
-            // Get the comments from the original definition of the containing symbol.
-            return containingSymbol.OriginalDefinition.GetDocumentationComment(compilation, expandIncludes: true, expandInheritdoc: true, cancellationToken: cancellationToken).GetParameterText(parameter.Name);
+            return formatter.Format(symbol, semanticModel, position, cancellationToken, CrefFormat)
+                ?? SpecializedCollections.EmptyEnumerable<TaggedText>();
```
do we need the `??` bit?  can we not just make `formatter.Format` always return a non-null value?

AdmiralSnyder replied:
probably. i wanted to keep the changes minimal

AdmiralSnyder replied:
done.

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -289,6 +348,33 @@ public string Format(string rawXmlText, Compilation compilation = null)
+            {
+                docSymbol = symbol.OriginalDefinition;
+            }
+            var rawXmlText = GetDocumentation(docSymbol, semanticModel.Compilation, cancellationToken);
```
```suggestion

            var rawXmlText = GetDocumentation(docSymbol, semanticModel.Compilation, cancellationToken);
```

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -289,6 +348,33 @@ public string Format(string rawXmlText, Compilation compilation = null)
+                docSymbol = symbol.OriginalDefinition;
+            }
+            var rawXmlText = GetDocumentation(docSymbol, semanticModel.Compilation, cancellationToken);
+
```
```suggestion
```

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -289,6 +348,33 @@ public string Format(string rawXmlText, Compilation compilation = null)
+            {
+                docSymbol = symbol.OriginalDefinition;
+            }
+            var rawXmlText = GetDocumentation(docSymbol, semanticModel.Compilation, cancellationToken);
```
```suggestion
            var rawXmlText = GetDocumentation(symbol.OriginalDefinition, semanticModel.Compilation, cancellationToken);
```

AdmiralSnyder replied:
i'm not sure here. basically, we need the special handling for type parameters only if there are any.
the special handling means: use the originaldefinition to get the documentation - right?

CyrusNajmabadi replied:
the code i wrote should be correct here.  we should always go back to the original-def to get teh documentation.

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -289,6 +348,33 @@ public string Format(string rawXmlText, Compilation compilation = null)
+
+            if (rawXmlText == null)
+            {
+                return null;
```
consider returning an empty enumerable here.

AdmiralSnyder replied:
this is how the other format methods defended against rawXmlText being null

CyrusNajmabadi replied:
the other format method can go away :)   it's not used anywhere after your change in this PR.

AdmiralSnyder replied:
done

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -527,6 +621,23 @@ private static void AppendTextFromAttribute(FormatterState state, XElement eleme
         }
 
+        internal static IEnumerable<SymbolDisplayPart> TypeParameterRefToSymbolDisplayParts(
+            string crefValue, ISymbol typeResolutionSymbol, int position, SemanticModel semanticModel, SymbolDisplayFormat format = null)
```
format should not be nullable/optional.

AdmiralSnyder replied:
it was in CrefToSymbolDisplayParts, so we need to change both, right?

CyrusNajmabadi replied:
you can if you want.  but this one should definitely change.

AdmiralSnyder replied:
done

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -527,6 +621,23 @@ private static void AppendTextFromAttribute(FormatterState state, XElement eleme
+            if (semanticModel != null)
+            {
+                var typeParameterIndex = typeResolutionSymbol.OriginalDefinition.GetAllTypeParameters().IndexOf(tp => tp.Name == crefValue);
+                if (typeParameterIndex > -1)
```
```suggestion
                if (typeParameterIndex >= 0)
```

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -527,6 +621,23 @@ private static void AppendTextFromAttribute(FormatterState state, XElement eleme
+                var typeParameterIndex = typeResolutionSymbol.OriginalDefinition.GetAllTypeParameters().IndexOf(tp => tp.Name == crefValue);
+                if (typeParameterIndex > -1)
+                {
+                    return typeResolutionSymbol.GetAllTypeArguments()[typeParameterIndex].ToMinimalDisplayParts(semanticModel, position, format);
```
need to make sure that typeParmeterIndex is within the range of type arguments returned.  

also, we need lots of tests for `<typeparamref ...` with cases of nested generics.

AdmiralSnyder replied:
do you mean unit tests by saying "lots of tests"? i haven't found any for these files.

CyrusNajmabadi replied:
yes, i'm saying unit tests.  @sharwell do you know where tests should go for this PR?

AdmiralSnyder replied:
added a check on GetAllTypeArguments().Length

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -527,6 +621,23 @@ private static void AppendTextFromAttribute(FormatterState state, XElement eleme
+
+            // if any of that fails fall back to just displaying the raw text
+            return SpecializedCollections.SingletonEnumerable(
+                new SymbolDisplayPart(SymbolDisplayPartKind.TypeParameterName, symbol: null, text: TrimCrefPrefix(crefValue)));
```
can youjust fall back to `CrefToSymbolDisplayParts` since htat what was called originally?

AdmiralSnyder replied:
i'd feel bad to do that, because it's not a cref

CyrusNajmabadi replied:
CrefToSymbolDisplayParts that used to be called for these values.  so this allows us to use the existing codepath for typeparams when they don't work with your new logic.

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/IDocumentationCommentFormattingService.cs:
```diff
@@ -11,5 +12,6 @@ internal interface IDocumentationCommentFormattingService : ILanguageService
     {
         string Format(string rawXmlText, Compilation compilation = null);
         IEnumerable<TaggedText> Format(string rawXmlText, SemanticModel semanticModel, int position, SymbolDisplayFormat format = null);
```
who uses thsi form?  is it still needed?

AdmiralSnyder replied:
Shift-F12 doesn't return (like, takes ages) so i cannot tell you.
but i remember that i got lots of errors when i wanted to change the method declaration, initially. that's why i decided to introduce another overload.
...
i modified the signature to get errors: there are 38 places that want that method.

CyrusNajmabadi replied:
there appear to be only one reference to this method.

AdmiralSnyder replied:
dunno why my last comment wasn't saved - i looked, and there were almost 40 places that referenced that overload.

CyrusNajmabadi replied:
there are not.  i checked.  this is the only location that references this.  if you have other locations you think reference this, can you please link to them?  Thanks!

AdmiralSnyder replied:
yeah, after intellisense stopped burning cpu, the reference count was visible, too
i was mistaken. overload is gone.

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/IDocumentationCommentFormattingService.cs:
```diff
@@ -11,5 +12,6 @@ internal interface IDocumentationCommentFormattingService : ILanguageService
     {
         string Format(string rawXmlText, Compilation compilation = null);
         IEnumerable<TaggedText> Format(string rawXmlText, SemanticModel semanticModel, int position, SymbolDisplayFormat format = null);
+        IEnumerable<TaggedText> Format(ISymbol symbol, SemanticModel semanticModel, int position, CancellationToken cancellationToken, SymbolDisplayFormat format = null);
```
CancellationToken should always be last.  Avoid optional parameters in internal interfaces (it's easy for callers to accidentally forget to pass a value).

AdmiralSnyder replied:
done

CyrusNajmabadi commented on src/Features/Core/Portable/QuickInfo/CommonSemanticQuickInfoProvider.cs:
```diff
@@ -288,13 +288,14 @@ void AddSection(string kind, ImmutableArray<TaggedText> taggedParts)
             else if (symbols.Any())
             {
-                var symbol = symbols.First().OriginalDefinition;
+                var symbol = symbols.First();//.OriginalDefinition;
```
```suggestion
                var symbol = symbols.First();
```

CyrusNajmabadi commented:
Overall looking good.  We'll need a bunch of tests here though.

AdmiralSnyder commented:
here are two screenshots i took:
![typeparamReplacement_Class](https://user-images.githubusercontent.com/16213022/74395032-866cb000-4e0e-11ea-9971-2a92b3f45665.png)
i hope it's ok that the text is german.
![TypeParamReplacement_Property](https://user-images.githubusercontent.com/16213022/74395033-879ddd00-4e0e-11ea-8ce8-fd637208a6e0.png)

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -289,6 +345,33 @@ public string Format(string rawXmlText, Compilation compilation = null)
+                docSymbol = symbol.OriginalDefinition;
+            }
+
+            var rawXmlText = GetDocumentation(docSymbol, semanticModel.Compilation, cancellationToken);
```
remove docSymbol entirely and just pass in `symbol.OriginalDefinition` here.

CyrusNajmabadi replied:
This comment still applies. This is just keeping the original logic that we used to have in the caller and moving it here.

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -289,6 +345,33 @@ public string Format(string rawXmlText, Compilation compilation = null)
+            return state.Builder;
+        }
+
         public IEnumerable<TaggedText> Format(string rawXmlText, SemanticModel semanticModel, int position, SymbolDisplayFormat format = null)
```
this overload should be removed.

AdmiralSnyder replied:
i did that.

CyrusNajmabadi commented:
Looks like QuickInfoForTypeParameterReference  failed. That's a good thing (since you changed behavuir).  Can you fixup that test and then add some more tests to validate the scenarios i mentioned above?

AdmiralSnyder commented:
i fixed the test, and added another checking on the method declaration to make sure that that still has the reference to T.
Is it okay if the test subs are named with a numbered prefix?
i.e. QuickInfoForTypeParameterReference and QuickInfoForTypeParameterReference2

CyrusNajmabadi commented:
> Is it okay if the test subs are named with a numbered prefix?

Yes, though discouraged. If you really can't make better names is ok. However, for these cases it does seem like better babes would be easy to find

CyrusNajmabadi commented:
@AdmiralSnyder does this still need work?

CyrusNajmabadi commented on src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:
```diff
@@ -257,6 +260,59 @@ private void EmitPendingChars()
+                default:
+                    return method.GetDocumentationComment(compilation, expandIncludes: true, expandInheritdoc: true, cancellationToken: cancellationToken).SummaryText;
+            }
+        }
```
why do we need these three methods?

AdmiralSnyder replied:
how should i answer that question?

AdmiralSnyder commented:
if i understand it correctly, the tests shouldn't fail in CI anymore. how does one make CI re-test without creating a new commit? - or, wait, do i have to merge something into my branch?

CyrusNajmabadi commented:
I'm getting this error on CI:

```
 src\EditorFeatures\Test2\IntelliSense\IntellisenseQuickInfoBuilderTests.vb#L1004

src\EditorFeatures\Test2\IntelliSense\IntellisenseQuickInfoBuilderTests.vb(1004,13): error BC30451: 'AssertEqualAdornments' is not declared. It may be inaccessible due to its protection level.

```

AdmiralSnyder commented:
@CyrusNajmabadi could you make the CI re-test, please?
seems the last commit didn't invoke a rebuild?

CyrusNajmabadi commented:
Thanks!
